File: presentation/src/main/java/com/mbmc/template/ui/RepoViewModel.java
package com.mbmc.template.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.mbmc.template.domain.entity.Repo;
import com.mbmc.template.domain.usecase.RepoUseCase;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import javax.inject.Inject;
import java.util.List;

public class RepoViewModel extends ViewModel {

    private static final String TAG = "REPO_VIEW_MODEL";

    private final MutableLiveData<DataWrapper<List<Repo>>> data = new MutableLiveData<>();
    private final CompositeDisposable disposables = new CompositeDisposable();
    private final RepoUseCase repoUseCase;

    @Inject
    public RepoViewModel(RepoUseCase repoUseCase) {
        this.repoUseCase = repoUseCase;
    }

    public LiveData<DataWrapper<List<Repo>>> observeRepos() {
        return data;
    }

    public void getRepos(String handle) {
        disposables.add(repoUseCase.getRepos(handle)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        repos -> data.setValue(new DataWrapper<>(repos)),
                        throwable -> data.setValue(new DataWrapper<>(throwable))
                ));
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposables.dispose();
    }

}

File: presentation/src/main/kotlin/com/mbmc/template/ui/RepoViewModel.kt
package com.mbmc.template.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.mbmc.template.domain.usecase.RepoUseCase
import com.mbmc.template.entity.Mapper
import com.mbmc.template.entity.Repo
import com.mbmc.template.util.Schedulers
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class RepoViewModel @Inject constructor(
    private val repoUseCase: RepoUseCase,
    private val mapper: Mapper,
    private val schedulers: Schedulers
) : ViewModel() {
    companion object {
        private const val TAG = "REPO_VIEW_MODEL"
    }

    private val _data = MutableLiveData<DataWrapper<List<Repo>>>()
    private val disposables = CompositeDisposable()

    val observeRepos: LiveData<DataWrapper<List<Repo>>>
        get() = _data

    fun getRepos(handle: String) {
        disposables.add(repoUseCase.getRepos(handle)
            .compose(schedulers.ioToMain())
            .doOnSubscribe {
                _data.value = DataWrapper(DataWrapper.State.LOADING)
            }
            .doFinally {
                _data.value = DataWrapper(DataWrapper.State.INIT)
            }
            .subscribe(
                { result -> _data.value = DataWrapper(result.map { data -> mapper.domainToPresentation(data) }) },
                { error -> _data.value = DataWrapper(error) }
            ))
    }

    override fun onCleared() {
        super.onCleared()
        disposables.dispose()
    }
}
